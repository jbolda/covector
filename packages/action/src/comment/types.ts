import { WebhookPayload } from "@actions/github/lib/interfaces";

interface PullRequestBranch {
  ref: string;
  repo: {
    url: string;
  };
  sha: string;
}

export interface PullRequestPayload extends WebhookPayload {
  pull_request: WebhookPayload["pull_request"] & {
    head: PullRequestBranch;
    base: PullRequestBranch;
  };
  repository: WebhookPayload["repository"] & {
    owner: {
      login: string;
    };
  };
}
