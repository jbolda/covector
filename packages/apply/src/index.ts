import {
  readPkgFile,
  writePkgFile,
  getPackageFileVersion,
  setPackageFileVersion,
  testSerializePkgFile,
} from "@covector/files";
import semver from "semver";
import { cloneDeep } from "lodash";

import type {
  PackageFile,
  ConfigFile,
  CommonBumps,
  ChangeParsed,
  Releases,
  PackageCommand,
} from "@covector/types";

export const apply = function* ({
  commands,
  config,
  cwd = process.cwd(),
  bump = true,
  previewVersion = "",
  prereleaseIdentifier = null,
}: {
  commands: PackageCommand[];
  config: ConfigFile;
  cwd: string;
  bump: boolean;
  previewVersion: string;
  prereleaseIdentifier: string | null;
}): Generator<any, PackageFile[], any> {
  const changes = commands.reduce(
    (finalChanges: { [k: string]: PackageCommand }, command) => {
      finalChanges[command.pkg] = command;
      return finalChanges;
    },
    {}
  );

  // @ts-ignore since TS doesn't like yielding on a Promise
  let allPackages = yield readAll({ changes, config, cwd });
  const bumps = bumpAll({
    changes,
    allPackages,
    previewVersion,
    prereleaseIdentifier,
  });

  if (bump) {
    yield writeAll({
      bumps: bumps.reduce(
        (final: PackageFile[], current) =>
          !current.vfile ? final : final.concat([current]),
        []
      ),
    });
  } else {
    bumps.forEach((b) => {
      if (!!b) console.log(`${b.name} planned to be bumped to ${b.version}`);
    });
  }
  return bumps;
};

export const validateApply = async ({
  commands,
  config,
  cwd = process.cwd(),
  prereleaseIdentifier = null,
}: {
  commands: PackageCommand[];
  config: ConfigFile;
  cwd: string;
  prereleaseIdentifier: string | null;
}) => {
  const changes = commands.reduce(
    (finalChanges: { [k: string]: PackageCommand }, command) => {
      finalChanges[command.pkg] = command;
      return finalChanges;
    },
    {}
  );
  let allPackages = await readAll({ changes, config, cwd });

  const bumps = bumpAll({
    changes,
    allPackages,
    logs: false,
    prereleaseIdentifier,
  }).reduce(
    (final: PackageFile[], current) =>
      !current.vfile ? final : final.concat([current]),
    []
  );

  try {
    for (let bump of bumps) {
      testSerializePkgFile({ packageFile: bump });
    }
    // will throw on validation error and not return true
    return true;
  } catch (e) {
    throw e;
  }
};

const readAll = async ({
  changes,
  config,
  cwd = process.cwd(),
}: {
  changes: { [k: string]: { parents: string[] } };
  config: ConfigFile;
  cwd: string;
}): Promise<{ [k: string]: PackageFile }> => {
  let templateShell: PackageFile = {
    version: "",
    pkg: { name: "", version: "" },
  };
  let files = Object.keys(changes).reduce(
    (fileList: { [k: string]: PackageFile }, change) => {
      fileList[change] = { ...templateShell };
      if (changes[change].parents && changes[change].parents.length > 0)
        changes[change].parents.forEach(
          (parent) => (fileList[parent] = { ...templateShell })
        );
      return fileList;
    },
    {}
  );

  const pkgs: string[] = Object.keys(files);
  const pkgFiles = await Promise.all(
    Object.keys(files).map((pkg) =>
      !config.packages[pkg].path
        ? { name: pkg }
        : readPkgFile({
            cwd,
            pkgConfig: config.packages[pkg],
            nickname: pkg,
          })
    )
  );

  return pkgs.reduce((list, pkg, index) => {
    list[pkg] = pkgFiles[index];
    return list;
  }, files);
};

const writeAll = function* ({ bumps }: { bumps: PackageFile[] }) {
  for (let bump of bumps) {
    yield writePkgFile({ packageFile: bump });
  }
};

const resolveParents = ({ config }: { config: ConfigFile }) => {
  return Object.keys(config.packages).reduce(
    (parents: { [k: string]: string[] }, pkg) => {
      parents[pkg] = [];
      Object.keys(config.packages).forEach((parent) => {
        if (
          !!config.packages[parent].dependencies &&
          config.packages[parent].dependencies!.includes(pkg)
        )
          parents[pkg].push(parent);
      });
      return parents;
    },
    {}
  );
};

type Changed = {
  [k: string]: {
    parents: string[];
    type: CommonBumps;
    changes?: ChangeParsed[];
  };
};
export const changesConsideringParents = ({
  assembledChanges,
  config,
  prereleaseIdentifier = null,
}: {
  assembledChanges: {
    releases: Releases;
    changes: ChangeParsed[];
  };
  config: ConfigFile;
  prereleaseIdentifier: string | null;
}) => {
  const parents = resolveParents({ config });

  let changes = Object.keys(assembledChanges.releases).reduce(
    (list: Changed, change) => {
      list[change] = assembledChanges.releases[change];
      list[change].parents = parents[change];
      return list;
    },
    {}
  );

  return {
    releases: parentBump(changes, parents, prereleaseIdentifier),
    changes: assembledChanges.changes,
  };
};

const parentBump = (
  initialChanges: Changed,
  parents: any,
  prereleaseIdentifier: string | null
): Changed => {
  let changes = { ...initialChanges };
  let recurse = false;
  Object.keys(initialChanges).forEach((main) => {
    if (changes[main].parents.length > 0) {
      changes[main].parents.forEach((pkg) => {
        // pkg is the parent and main is the child
        if (!!changes[pkg]) {
          // if a change is planned on the parent
          // we don't need to plan a release
        } else {
          // if the parent doesn't have a release
          // add one to adopt the next version of it's child
          changes[pkg] = {
            ...cloneDeep(changes[main]),
            // prerelease will do bump the X in `-beta.X` if it is already a prerelease
            // or it will do a prepatch if it isn't a prerelease
            type: !prereleaseIdentifier ? "patch" : "prerelease",
          };
          if (changes[pkg].changes) {
            changes[pkg].changes!.forEach((parentChange) => {
              parentChange.meta.dependencies = `Bumped due to a bump in ${main}.`;
            });
          }
          changes[pkg].parents = parents[pkg];
          // we also need to presume recursion to update the parents' parents
          recurse = true;
        }
      });
    }
  });
  return recurse ? parentBump(changes, parents, prereleaseIdentifier) : changes;
};

const bumpAll = ({
  changes,
  allPackages,
  logs = true,
  previewVersion = "",
  prereleaseIdentifier = null,
}: {
  changes: Releases;
  allPackages: { [k: string]: PackageFile };
  logs?: boolean;
  previewVersion?: string;
  prereleaseIdentifier: string | null;
}) => {
  let packageFiles = { ...allPackages };
  for (let pkg of Object.keys(changes)) {
    if (!packageFiles[pkg].vfile || changes[pkg].type === "noop") continue;
    if (logs && !previewVersion)
      console.log(`bumping ${pkg} with ${changes[pkg].type}`);
    if (previewVersion)
      // change log (assume that the prerelease will be removed)
      console.log(
        `bumping ${pkg} with ${previewVersion} template to publish a preview`
      );
    packageFiles[pkg] = bumpMain({
      packageFile: packageFiles[pkg],
      bumpType: changes[pkg].type,
      previewVersion,
      prereleaseIdentifier,
      errorOnVersionRange: changes[pkg].errorOnVersionRange,
    });
    if (changes[pkg] && changes[pkg].dependencies) {
      let deps = changes[pkg].dependencies!;
      for (let pkgDep of deps) {
        if (!!changes[pkgDep]) {
          packageFiles[pkg] = bumpDeps({
            packageFile: packageFiles[pkg],
            dep: pkgDep,
            bumpType: changes[pkgDep].type,
            previewVersion,
            prereleaseIdentifier,
          });
        }
      }
    }
  }

  return Object.keys(packageFiles).map((pkg) => packageFiles[pkg]);
};

const bumpMain = ({
  packageFile,
  bumpType,
  previewVersion,
  prereleaseIdentifier = null,
  errorOnVersionRange,
}: {
  packageFile: PackageFile;
  bumpType: CommonBumps;
  previewVersion: string;
  prereleaseIdentifier: string | null;
  errorOnVersionRange?: string;
}) => {
  let pkg = { ...packageFile };
  if (!pkg.version)
    throw new Error(`${pkg.name} does not have a version number.`);
  // @ts-ignore TODO bumpType should be narrowed to meet ReleaseType
  let next = semver.inc(pkg.version, bumpType, prereleaseIdentifier);
  if (next) {
    pkg.version = next;
    pkg.versionMajor = semver.major(next);
    pkg.versionMinor = semver.minor(next);
    pkg.versionPatch = semver.patch(next);
    pkg.versionPrerelease = semver.prerelease(next);
  }
  const prevVersion = getPackageFileVersion({ pkg });
  const preVersionCleaned = semver.prerelease(prevVersion) ? semver.inc(prevVersion, 'patch') : prevVersion;
  let version =
    previewVersion && previewVersion !== ""
      ? semver.valid(`${preVersionCleaned}-${previewVersion}`)
      : // @ts-ignore TODO bumpType should be narrowed to meet ReleaseType
        semver.inc(prevVersion, bumpType, prereleaseIdentifier);
  if (version) {
    pkg = setPackageFileVersion({ pkg, version });
    if (errorOnVersionRange && semver.satisfies(version, errorOnVersionRange)) {
      throw new Error(
        `${pkg.name} will be bumped to ${version}. ` +
          `This satisfies the range ${errorOnVersionRange} which the configuration disallows. ` +
          `Please adjust your bump to accommodate the range or otherwise adjust the allowed range in \`errorOnVersionRange\`.`
      );
    }
  }
  return pkg;
};

const bumpDeps = ({
  packageFile,
  dep,
  bumpType,
  previewVersion,
  prereleaseIdentifier = null,
}: {
  packageFile: PackageFile;
  dep: string;
  bumpType: string;
  previewVersion: string;
  prereleaseIdentifier: string | null;
}) => {
  let pkg = { ...packageFile };

  if (pkg.pkg && pkg.vfile)
    ["dependencies", "devDependencies", "dev-dependencies"].forEach(
      (property) => {
        // @ts-ignore
        if (pkg.pkg[property]) {
          // @ts-ignore
          Object.keys(pkg.pkg[property]).forEach((existingDep) => {
            if (existingDep === dep) {
              const prevVersion = getPackageFileVersion({ pkg, property, dep });
              const preVersionCleaned = semver.prerelease(prevVersion) ? semver.inc(prevVersion, 'patch') : prevVersion;
              let version =
                previewVersion && previewVersion !== ""
                  ? semver.valid(`${preVersionCleaned}-${previewVersion}`)
                  : incConsideringPartials(
                      dep,
                      prevVersion,
                      // @ts-ignore TODO deal with ReleaseType
                      bumpType,
                      prereleaseIdentifier
                    );
              if (version)
                pkg = setPackageFileVersion({ pkg, version, property, dep });
            }
          });
        }
      }
    );

  return pkg;
};

const incConsideringPartials = (
  dependency: string,
  version: string,
  bumpType: string,
  prereleaseIdentifier: string | null
) => {
  if (semver.valid(version)) {
    // @ts-ignore TODO deal with ReleaseType
    return semver.inc(version, bumpType, prereleaseIdentifier);
  } else {
    if (prereleaseIdentifier !== null) {
      console.warn(
        `bump for ${dependency} skipped as ${version} is a range, and does not specifically include prereleases. ` +
          `Please pin to a major.minor.patch for a prerelease bump.`
      );
      return null;
    }
    try {
      const coerced = semver.coerce(version);
      if (!coerced)
        throw new Error(
          `Cannot bump ${version} with ${bumpType}. Is it a valid version number?`
        );
      // @ts-ignore TODO deal with ReleaseType
      const fullVersion = semver.inc(coerced, bumpType).split(".");
      if (version.split(".").length === 2) {
        return [fullVersion[0], fullVersion[1]].join(".");
      } else if (version.split(".").length === 1) {
        return fullVersion[0];
      } else {
        // failsafe is better than null
        return fullVersion.join(".");
      }
    } catch {
      // failsafe is better than null
      // @ts-ignore TODO what should we _really_ do here?
      return semver.inc(semver.coerce(version), bumpType);
    }
  }
};
