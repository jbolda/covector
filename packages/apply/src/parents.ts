import type {
  ConfigFile,
  Changed,
  ChangeParsed,
  Releases,
  PackageFile,
  DepsKeyed,
} from "@covector/types";

const resolveParents = ({
  config,
  allPackages,
}: {
  config: ConfigFile;
  allPackages?: Record<string, PackageFile>;
}) => {
  return Object.keys(config.packages).reduce(
    (parents: Record<string, Record<string, DepsKeyed>>, pkg) => {
      parents[pkg] = {};
      Object.keys(config.packages).forEach((parent) => {
        if (
          !!config.packages[parent].dependencies &&
          (config.packages[parent].dependencies ?? []).includes(pkg)
        ) {
          parents[pkg][parent] = !allPackages ? {} : allPackages?.[parent].deps;
        }
      });
      return parents;
    },
    {}
  );
};

export const changesConsideringParents = ({
  assembledChanges,
  config,
  allPackages,
  prereleaseIdentifier,
}: {
  assembledChanges: {
    releases: Releases;
    changes: ChangeParsed[];
  };
  config: ConfigFile;
  allPackages: Record<string, PackageFile>;
  prereleaseIdentifier?: string;
}) => {
  const parents = resolveParents({ config, allPackages });

  let initialChanges = Object.keys(assembledChanges.releases).reduce(
    (list: Changed, change) => {
      list[change] = assembledChanges.releases[change];
      list[change].parents = parents[change];
      return list;
    },
    {}
  );

  const releases = parentBump({
    initialChanges,
    parents,
    prereleaseIdentifier,
  });

  return {
    releases,
    changes: assembledChanges.changes,
  };
};

const parentBump = ({
  initialChanges,
  parents,
  prereleaseIdentifier,
}: {
  initialChanges: Changed;
  parents: any;
  prereleaseIdentifier?: string;
}): Changed => {
  let changes = { ...initialChanges };
  let recurse = false;
  Object.keys(initialChanges).forEach((main) => {
    if (Object.keys(changes[main].parents).length > 0) {
      Object.entries(changes[main].parents).forEach(([pkg, deps]) => {
        // if we can't find this package in the dependencies, mark it as none
        //   and presume that the intent is for it to receive a patch bump from a dep
        const prevDepVersion =
          deps?.[main]?.find((depTypes) => depTypes.type === "dependencies")
            ?.version || "none";
        const versionRequirementMatch = /[\^=~]/.exec(prevDepVersion);
        // pkg is the parent and main is the child
        if (
          !changes[pkg] &&
          (!versionRequirementMatch || prevDepVersion === "none")
        ) {
          // if the parent doesn't have a release
          //   and it doesn't have the dependency as a range
          //   add one to adopt the next version of it's child
          //   but don't use the changes as we will refer to this dep
          //   bump otherwise and pulling up the changelog is not needed
          changes[pkg] = {
            changes: [],
            parents: parents[pkg],
            // prerelease will do bump the X in `-beta.X` if it is already a prerelease
            // or it will do a prepatch if it isn't a prerelease
            type: !prereleaseIdentifier ? "patch" : "prerelease",
          };
          // we also need to presume recursion to update the parents' parents
          if (Object.values(parents[pkg]).length > 0) recurse = true;

          changes[pkg].changes?.forEach((parentChange) => {
            // this ends up overwriting in cases multiple bumps,
            //   we should adjust this to accept multiple
            if (
              !parentChange.meta.dependencies ||
              parentChange.meta.dependencies.length === 0
            ) {
              parentChange.meta.dependencies = [main];
            } else {
              parentChange.meta.dependencies.push(main);
            }
          });
        }
      });
    }
  });
  return recurse
    ? parentBump({ initialChanges: changes, parents, prereleaseIdentifier })
    : changes;
};
