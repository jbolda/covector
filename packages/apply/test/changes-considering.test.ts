import { changesConsideringParents } from "../src";
import { PackageFile, CommonBumps } from "@covector/types";
import type { ConfigFile, PackageConfig } from "@covector/types";
import { describe, it } from "../../../helpers/test-scope.ts";
import { expect } from "vitest";

const allPackagesWithoutRead = ({
  config,
  pkgDepVersion,
}: {
  config: ConfigFile;
  pkgDepVersion?: Record<string, Record<string, string>>;
}) =>
  Object.entries(config.packages)
    .map(
      ([pkgName, configInfo]: [pkgName: string, configInfo: PackageConfig]) => {
        return {
          name: pkgName,
          version: "none",
          deps: (configInfo.dependencies ?? []).reduce(
            (deps, dep) => {
              const version = pkgDepVersion?.[pkgName]?.[dep]
                ? pkgDepVersion[pkgName][dep]
                : "none";
              deps[dep] = [{ type: "dependencies", version }];
              return deps;
            },
            {} as Record<string, { type: "dependencies"; version: string }[]>
          ),
        };
      }
    )
    .reduce(
      (pkgs, pkg: any) => {
        if (pkg.name) pkgs[pkg.name] = pkg;
        return pkgs;
      },
      {} as Record<string, PackageFile>
    );

describe("list changes considering parents", () => {
  it("adds changes for dependency", function* () {
    const assembledChanges = {
      releases: {
        all: {
          dependencies: undefined,
          manager: "javascript",
          path: undefined,
          pkg: "all",
          type: "minor",
        },
      },
    };

    const config = {
      packages: {
        "yarn-workspace-base-pkg-a": {
          path: "./packages/pkg-a/",
          manager: "javascript",
          dependencies: ["yarn-workspace-base-pkg-b", "all"],
        },
        "yarn-workspace-base-pkg-b": {
          path: "./packages/pkg-b/",
          manager: "javascript",
          dependencies: ["all"],
        },
        all: { version: true },
      },
    };

    //@ts-expect-error
    const changes = changesConsideringParents({ assembledChanges, config });

    expect({
      changes,
    }).toMatchSnapshot();
  });

  it("bumps patch due to dependency bump", function* () {
    const assembledChanges = {
      releases: {
        "yarn-workspace-base-pkg-a": {
          dependencies: undefined,
          manager: "javascript",
          path: undefined,
          pkg: "all",
          type: "patch",
        },
        all: {
          dependencies: undefined,
          manager: "javascript",
          path: undefined,
          pkg: "all",
          type: "minor",
        },
      },
    };

    const config = {
      packages: {
        "yarn-workspace-base-pkg-a": {
          path: "./packages/pkg-a/",
          manager: "javascript",
          dependencies: ["yarn-workspace-base-pkg-b", "all"],
        },
        "yarn-workspace-base-pkg-b": {
          path: "./packages/pkg-b/",
          manager: "javascript",
          dependencies: ["all"],
        },
        all: { version: true },
      },
    };

    //@ts-expect-error
    const changes = changesConsideringParents({ assembledChanges, config });

    expect({
      changes,
    }).toMatchSnapshot();
  });

  it("rolls up the parent bumps", function* () {
    const changesFiles = [
      {
        releases: {
          "pkg-c": "patch",
          "pkg-d": "major",
        },
        summary:
          "This should patch bump up the pkg-[number] line to where pkg-one also receives a bump. The pkg-c writes to a patch so we patch bump up that line too.",
        meta: {},
      },
      {
        releases: {
          "pkg-b": "minor",
        },
        summary:
          "The pkg-b doesn't have a dep bump, but can dep bump pkg-a and pkg-overall with a patch.",
        meta: {},
      },
    ];

    const assembledChanges = {
      releases: {
        "pkg-b": {
          type: "minor",
          changes: [changesFiles[1]],
        },
        "pkg-c": {
          type: "patch",
          changes: [changesFiles[0]],
        },
        "pkg-d": {
          type: "major",
          changes: [changesFiles[0]],
        },
      },
    };

    const config = {
      changeFolder: ".changes",
      packages: {
        "pkg-overall": {
          path: "./packages/pkg-overall/",
          manager: "javascript",
          dependencies: ["pkg-a", "pkg-b"],
        },
        "pkg-a": {
          path: "./packages/pkg-a/",
          manager: "javascript",
          dependencies: ["pkg-c"],
        },
        "pkg-b": {
          path: "./packages/pkg-b/",
          manager: "javascript",
          dependencies: [],
        },
        "pkg-c": {
          path: "./packages/pkg-c/",
          manager: "javascript",
          dependencies: ["pkg-d"],
        },
        "pkg-d": {
          path: "./packages/pkg-d/",
          manager: "javascript",
          dependencies: [],
        },
        "pkg-one": {
          path: "./packages/pkg-one/",
          manager: "javascript",
          dependencies: ["pkg-two"],
        },
        "pkg-two": {
          path: "./packages/pkg-two/",
          manager: "javascript",
          dependencies: ["pkg-three"],
        },
        "pkg-three": {
          path: "./packages/pkg-three/",
          manager: "javascript",
          dependencies: ["pkg-four"],
        },
        "pkg-four": {
          path: "./packages/pkg-four/",
          manager: "javascript",
          dependencies: ["pkg-d"],
        },
      },
    };

    const allPackages = allPackagesWithoutRead({ config });

    const changes = changesConsideringParents({
      //@ts-expect-error
      assembledChanges,
      config,
      allPackages,
    });
    // console.error(changes)

    // these are directly defined in the change files
    expect(changes.releases["pkg-b"].type).toBe("minor");
    expect(changes.releases["pkg-c"].type).toBe("patch");
    expect(changes.releases["pkg-d"].type).toBe("major");

    // these are the top level rolled up
    expect(changes.releases["pkg-overall"].type).toBe("patch");
    expect(changes.releases["pkg-a"].type).toBe("patch");

    // rolls up from pkg-d
    expect(changes.releases["pkg-four"].type).toBe("patch");
    expect(changes.releases["pkg-three"].type).toBe("patch");
    expect(changes.releases["pkg-two"].type).toBe("patch");
    expect(changes.releases["pkg-one"].type).toBe("patch");

    expect({
      changes,
    }).toMatchSnapshot();
  });

  it("skips roll up if range dep", function* () {
    const changesFiles = [
      {
        releases: {
          "pkg-a": "patch",
        },
        summary: "Roll up just a bit please.",
        meta: {},
      },
    ];

    // this is the initial set of changes
    const assembledChanges = {
      releases: {
        "pkg-a": {
          type: "patch" as CommonBumps,
          changes: [changesFiles[0]],
        },
      },
    };

    const config = {
      changeFolder: ".changes",
      packages: {
        "pkg-a": {
          path: "./packages/pkg-a/",
          manager: "javascript",
          dependencies: [],
        },
        "pkg-b": {
          path: "./packages/pkg-b/",
          manager: "javascript",
          dependencies: ["pkg-a"],
        },
        "pkg-c": {
          path: "./packages/pkg-c/",
          manager: "javascript",
          dependencies: ["pkg-a"],
        },
      },
    };

    const pkgDepVersion = {
      "pkg-b": { "pkg-a": "1.0.0" },
      "pkg-c": { "pkg-a": "^1.0.0" },
    };
    const allPackages = allPackagesWithoutRead({ config, pkgDepVersion });

    const changes = changesConsideringParents({
      //@ts-expect-error
      assembledChanges,
      config,
      allPackages,
    });
    // console.error(changes)

    // these are directly defined in the change files
    expect(changes.releases["pkg-a"].type).toBe("patch");

    // these are the top level rolled up
    expect(changes.releases["pkg-b"].type).toBe("patch");

    // rolls up from pkg-d
    expect(changes.releases["pkg-c"]).toBeUndefined();

    expect({
      changes,
    }).toMatchSnapshot();
  });
});
