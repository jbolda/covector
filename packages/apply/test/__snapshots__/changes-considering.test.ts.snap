// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list changes considering parents > adds changes for dependency 1`] = `
{
  "changes": {
    "changes": undefined,
    "releases": {
      "all": {
        "dependencies": undefined,
        "manager": "javascript",
        "parents": {
          "yarn-workspace-base-pkg-a": {},
          "yarn-workspace-base-pkg-b": {},
        },
        "path": undefined,
        "pkg": "all",
        "type": "minor",
      },
      "yarn-workspace-base-pkg-a": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "all",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "all",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "yarn-workspace-base-pkg-b",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {},
        "type": "patch",
      },
      "yarn-workspace-base-pkg-b": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "all",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "all",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {
          "yarn-workspace-base-pkg-a": {},
        },
        "type": "patch",
      },
    },
  },
}
`;

exports[`list changes considering parents > bumps patch due to dependency bump 1`] = `
{
  "changes": {
    "changes": undefined,
    "releases": {
      "all": {
        "dependencies": undefined,
        "manager": "javascript",
        "parents": {
          "yarn-workspace-base-pkg-a": {},
          "yarn-workspace-base-pkg-b": {},
        },
        "path": undefined,
        "pkg": "all",
        "type": "minor",
      },
      "yarn-workspace-base-pkg-a": {
        "dependencies": undefined,
        "manager": "javascript",
        "parents": {},
        "path": undefined,
        "pkg": "all",
        "type": "patch",
      },
      "yarn-workspace-base-pkg-b": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "all",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "all",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {
          "yarn-workspace-base-pkg-a": {},
        },
        "type": "patch",
      },
    },
  },
}
`;

exports[`list changes considering parents > rolls up the parent bumps 1`] = `
{
  "changes": {
    "changes": undefined,
    "releases": {
      "pkg-a": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "pkg-c",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-c",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-c",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-c",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {
          "pkg-overall": {
            "pkg-a": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
            "pkg-b": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
        },
        "type": "patch",
      },
      "pkg-b": {
        "changes": [
          {
            "meta": {},
            "releases": {
              "pkg-b": "minor",
            },
            "summary": "The pkg-b doesn't have a dep bump, but can dep bump pkg-a and pkg-overall with a patch.",
          },
        ],
        "parents": {
          "pkg-overall": {
            "pkg-a": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
            "pkg-b": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
        },
        "type": "minor",
      },
      "pkg-c": {
        "changes": [
          {
            "meta": {},
            "releases": {
              "pkg-c": "patch",
              "pkg-d": "major",
            },
            "summary": "This should patch bump up the pkg-[number] line to where pkg-one also receives a bump. The pkg-c writes to a patch so we patch bump up that line too.",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {
          "pkg-a": {
            "pkg-c": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
        },
        "type": "patch",
      },
      "pkg-d": {
        "changes": [
          {
            "meta": {},
            "releases": {
              "pkg-c": "patch",
              "pkg-d": "major",
            },
            "summary": "This should patch bump up the pkg-[number] line to where pkg-one also receives a bump. The pkg-c writes to a patch so we patch bump up that line too.",
          },
        ],
        "parents": {
          "pkg-c": {
            "pkg-d": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
          "pkg-four": {
            "pkg-d": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
        },
        "type": "major",
      },
      "pkg-four": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-d",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {
          "pkg-three": {
            "pkg-four": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
        },
        "type": "patch",
      },
      "pkg-one": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "pkg-two",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {},
        "type": "patch",
      },
      "pkg-overall": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "pkg-b",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-b",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-a",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-b",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-a",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-b",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-a",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {},
        "type": "patch",
      },
      "pkg-three": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "pkg-four",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-four",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-four",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {
          "pkg-two": {
            "pkg-three": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
        },
        "type": "patch",
      },
      "pkg-two": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "pkg-three",
              ],
            },
            "releases": {},
            "summary": "",
          },
          {
            "meta": {
              "dependencies": [
                "pkg-three",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {
          "pkg-one": {
            "pkg-two": [
              {
                "type": "dependencies",
                "version": "none",
              },
            ],
          },
        },
        "type": "patch",
      },
    },
  },
}
`;

exports[`list changes considering parents > skips roll up if range dep 1`] = `
{
  "changes": {
    "changes": undefined,
    "releases": {
      "pkg-a": {
        "changes": [
          {
            "meta": {},
            "releases": {
              "pkg-a": "patch",
            },
            "summary": "Roll up just a bit please.",
          },
        ],
        "parents": {
          "pkg-b": {
            "pkg-a": [
              {
                "type": "dependencies",
                "version": "1.0.0",
              },
            ],
          },
          "pkg-c": {
            "pkg-a": [
              {
                "type": "dependencies",
                "version": "^1.0.0",
              },
            ],
          },
        },
        "type": "patch",
      },
      "pkg-b": {
        "changes": [
          {
            "meta": {
              "dependencies": [
                "pkg-a",
              ],
            },
            "releases": {},
            "summary": "",
          },
        ],
        "parents": {},
        "type": "patch",
      },
    },
  },
}
`;
